File: Python/Tree/Segment Tree/315. Count of Smaller Numbers After Self.py
class Solution:
    
    def update(self, index, val, tree, size):
        index += size
        tree[index] += val
        while index > 1:
            index = index // 2
            tree[index] = tree[index * 2] + tree[index * 2 + 1]
    
    def query(self, left, right, tree, size):
        ans = 0
        left += size
        right += size
        while left < right:
            if left % 2 == 1:
                ans += tree[left]
                left += 1
            left = left // 2
            if right % 2 == 1:
                right -= 1
                ans += tree[right]
            right = right // 2
        return ans
    
    def countSmaller(self, nums):
        offset = 10000
        size = 20001
        tree = [0] * (size * 2)
        ans = []
        for i in range(len(nums)-1, -1, -1):
            ans.append(self.query(0, nums[i] + offset, tree, size))
            self.update(nums[i] + offset, 1, tree, size)
        return ans[::-1]

File: C++/Tree/Segment Tree/315. Count of Smaller Numbers After Self.cpp
class Solution {
    
    int offset = 10000;
    int n = 20001;
    vector<int> tree;
    
public:
    
    void updateTree(int node, int left, int right, int index) {
        if (left == right) {
            tree[node]++;
        }
        else {
            int mid = left + (right - left) / 2;
            if (index <= mid) {
                updateTree(node * 2, left, mid, index);
            }
            else {
                updateTree(node * 2 + 1, mid + 1, right, index);
            }
            tree[node] = tree[node * 2] + tree[node * 2 + 1];
        }
    }
    
    int queryTree(int node, int left, int right, int queryLeft, int queryRight) {
        if (queryRight < left or right < queryLeft) {
            return 0;
        }
        else if (queryLeft <= left and right <= queryRight) {
            return tree[node];
        }
        else {
            int mid = left + (right - left) / 2;
            int q1 = queryTree(node * 2, left, mid, queryLeft, queryRight);
            int q2 = queryTree(node * 2 + 1, mid + 1, right, queryLeft, queryRight);
            return q1 + q2;
        }
    }
    
    vector<int> countSmaller(vector<int>& nums) {
        tree.resize(4 * n, 0);
        vector<int> ans(nums.size());
        for (int i = nums.size() - 1; i >= 0; i--) {
            int num = nums[i] + offset;
            ans[i] = queryTree(1, 0, 20000, 0, num - 1);
            updateTree(1, 0, 20000, num);
        }
        return ans;
    }
};